File: gaia-sdk-python/gaia_sdk/gaia.py
from typing import Callable, List, Union

import requests
from rx.core.typing import Observable
from rx.scheduler import ThreadPoolScheduler

from gaia_sdk.api import ISensorStream, IdentityOp
from gaia_sdk.api.GaiaCredentials import UsernamePasswordCredentials, GaiaCredentials, JWTCredentials
from gaia_sdk.api.ISensorFunction import ISensorFunction
from gaia_sdk.api.SkillRef import SkillRef
from gaia_sdk.api.data.DataRef import DataRef
from gaia_sdk.graphql import RetrievalReq, ExperienceReq, KnowledgeReq, EdgeReq, \
    IntentReq, IdentityReq, PromptReq, StatementReq, FulfilmentReq, CodeReq, BehaviourReq, IntrospectionReq, \
    PerceptionReq, PreservationReq, CreatedIdentityImpulse, UpdatedIdentityImpulse, DeletedIdentityImpulse, \
    CreatedIntentImpulse, UpdatedIntentImpulse, DeletedIntentImpulse, RetrievalRes, \
    ExperienceRes, EdgeRes, StatementRes, PromptRes, IntentRes, IdentityRes, KnowledgeRes, FulfilmentRes, CodeRes, \
    BehaviourRes, IntrospectionRes, PreservationRes, PerceptionRes, PerceivedImpulse, \
    PerceiveDataImpulse, PerceiveActionImpulse, DeleteIdentityImpulse, UpdateIdentityImpulse, CreateIdentityImpulse, \
    DeleteIntentImpulse, UpdateIntentImpulse, CreateIntentImpulse, \
    CreatePromptImpulse, UpdatePromptImpulse, DeletePromptImpulse, CreatedPromptImpulse, UpdatedPromptImpulse, \
    DeletedPromptImpulse, CreateStatementImpulse, UpdateStatementImpulse, DeleteStatementImpulse, \
    CreatedStatementImpulse, UpdatedStatementImpulse, DeletedStatementImpulse, CreateFulfilmentImpulse, \
    UpdateFulfilmentImpulse, DeleteFulfilmentImpulse, CreatedFulfilmentImpulse, UpdatedFulfilmentImpulse, \
    DeletedFulfilmentImpulse, CreateBehaviourImpulse, UpdateBehaviourImpulse, DeleteBehaviourImpulse, \
    CreatedBehaviourImpulse, UpdatedBehaviourImpulse, DeletedBehaviourImpulse, CreateCodeImpulse, UpdateCodeImpulse, \
    DeleteCodeImpulse, CreatedCodeImpulse, UpdatedCodeImpulse, DeletedCodeImpulse, CreateEdgeImpulse, \
    DeleteEdgeImpulse, CreatedEdgeImpulse, DeletedEdgeImpulse, BehaviourExecutionRes, BehaviourExecutionReq, \
    BehaviourExecutionDetailReq, BehaviourExecutionDetailRes, \
    CreatedSkillBuildJobImpulseRes, CreatedSkillBuildJobImpulseReq, \
    CancelSkillBuildJobImpulse, CreateSkillBuildJobImpulse, CanceledSkillBuildJobImpulseRes, \
    CanceledSkillBuildJobImpulseReq, \
    PracticeReq, PracticeRes, SkillBuildJobRes, SkillBuildJobReq
from gaia_sdk.graphql.GaiaClientFactory import GaiaClientFactory
from gaia_sdk.http.GaiaStreamClientBuilder import GaiaStreamClientFactory
from gaia_sdk.http.HttpSensorFunction import HttpSensorFunction
from gaia_sdk.http.HttpSensorStream import HttpSensorStream
from gaia_sdk.http.response.LoggedIn import LoggedIn

Uuid = str


class Gaia:
    _pool = ThreadPoolScheduler(5)
    _client_factory = GaiaClientFactory()
    _stream_client_factory = GaiaStreamClientFactory()

    @classmethod
    def connect(cls, url: str, credentials: GaiaCredentials) -> 'GaiaRef':
        config = GaiaConfig(url, HttpSensorFunction(url, credentials, cls._pool, cls._client_factory),
                            HttpSensorStream(url, credentials, cls._pool, cls._stream_client_factory))
        return GaiaRef(config, config.functionProcessor, config.streamProcessor)

    @classmethod
    def login(cls, url: str, credentials: UsernamePasswordCredentials) -> 'GaiaRef':
        headers = {'Content-Type': 'application/json'}
        response = requests.post(f"{url}/api/auth/access", json=credentials.__repr__(), headers=headers)
        response.raise_for_status()
        return Gaia.connect(url, JWTCredentials(LoggedIn(response.json()).access_token))


class GaiaConfig:
    url: str
    functionProcessor: ISensorFunction
    streamProcessor: ISensorStream

    def __init__(self, url: str, functionProcessor: ISensorFunction, streamProcessor: ISensorStream):
        self.url = url
        self.functionProcessor = functionProcessor
        self.streamProcessor = streamProcessor

    def __eq__(self, other):
        return self.url == other.url

    def __repr__(self):
        return {'url': self.url}


class GaiaRef(ISensorFunction):  # TODO: implement ISensorStream
    config: GaiaConfig
    f_proc: ISensorFunction
    s_proc: ISensorStream

    def __init__(self, config: GaiaConfig, f_proc: ISensorFunction, s_proc: ISensorStream):
        self.config = config
        self.f_proc = f_proc
        self.s_proc = s_proc

    def __eq__(self, other):
        return self.config == other.config

    def __repr__(self):
        return {'config': self.config}

    def data(self, uri: str) -> DataRef:
        return self.s_proc.data(uri)

    def identity(self) -> IdentityOp:
        return self.s_proc.identity()

    def skill(self, url: str) -> SkillRef:
        return self.s_proc.skill(url)

    def retrieve(self, config: Callable[[RetrievalReq], None]) -> Observable[RetrievalRes]:
        return self.f_proc.retrieve(config)

    def retrieve_experience(self, config: Callable[[ExperienceReq], None]) -> Observable[ExperienceRes]:
        return self.f_proc.retrieve_experience(config)

    def retrieve_knowledge(self, config: Callable[[KnowledgeReq], None]) -> Observable[KnowledgeRes]:
        return self.f_proc.retrieve_knowledge(config)

    def retrieve_edges(self, source: Uuid, config: Callable[[EdgeReq], None], limit: int = None, offset: int = None) -> \
            Observable[EdgeRes]:
        return self.f_proc.retrieve_edges(source, config, limit, offset)

    def retrieve_edge(self, source: Uuid, edgeId: Uuid, config: Callable[[EdgeReq], None]) -> Observable[EdgeRes]:
        return self.f_proc.retrieve_edge(source, edgeId, config)

    def retrieve_identities(self, config: Callable[[IdentityReq], None], limit: int = None, offset: int = None) -> \
            Observable[IdentityRes]:
        return self.f_proc.retrieve_identities(config, limit, offset)

    def retrieve_identity(self, identityId: Uuid, config: Callable[[IdentityReq], None]) -> Observable[IdentityRes]:
        return self.f_proc.retrieve_identity(identityId, config)

    def retrieve_intents(self, identityId: Uuid, config: Callable[[IntentReq], None], limit: int = None,
                         offset: int = None) -> Observable[IntentRes]:
        return self.f_proc.retrieve_intents(identityId, config, limit, offset)

    def retrieve_intent(self, identityId: Uuid, reference: Uuid, config: Callable[[IntentReq], None]) -> Observable[
        IntentRes]:
        return self.f_proc.retrieve_intent(identityId, reference, config)

    def retrieve_prompts(self, identityId: Uuid, config: Callable[[PromptReq], None], limit: int = None,
                         offset: int = None) -> Observable[PromptRes]:
        return self.f_proc.retrieve_prompts(identityId, config, limit, offset)

    def retrieve_prompt(self, identityId: Uuid, reference: Uuid, config: Callable[[PromptReq], None]) -> Observable[
        PromptRes]:
        return self.f_proc.retrieve_prompt(identityId, reference, config)

    def retrieve_statements(self, identityId: Uuid, config: Callable[[StatementReq], None], limit: int = None,
                            offset: int = None) -> Observable[StatementRes]:
        return self.f_proc.retrieve_statements(identityId, config, limit, offset)

    def retrieve_statement(self, identityId: Uuid, reference: Uuid, config: Callable[[StatementReq], None]) -> \
            Observable[StatementRes]:
        return self.f_proc.retrieve_statement(identityId, reference, config)

    def retrieve_fulfilments(self, identityId: Uuid, config: Callable[[FulfilmentReq], None], limit: int = None,
                             offset: int = None) -> Observable[FulfilmentRes]:
        return self.f_proc.retrieve_fulfilments(identityId, config, limit, offset)

    def retrieve_fulfilment(self, identityId: Uuid, reference: Uuid, config: Callable[[FulfilmentReq], None]) -> \
            Observable[FulfilmentRes]:
        return self.f_proc.retrieve_fulfilment(identityId, reference, config)

    def retrieve_codes(self, identityId: Uuid, config: Callable[[CodeReq], None], limit: int = None,
                       offset: int = None) -> Observable[CodeRes]:
        return self.f_proc.retrieve_codes(identityId, config, limit, offset)

    def retrieve_code(self, identityId: Uuid, reference: Uuid, config: Callable[[CodeReq], None]) -> Observable[
        CodeRes]:
        return self.f_proc.retrieve_code(identityId, reference, config)

    def retrieve_behaviours(self, identityId: Uuid, config: Callable[[BehaviourReq], None], limit: int = None,
                            offset: int = None) -> Observable[BehaviourRes]:
        return self.f_proc.retrieve_behaviours(identityId, config, limit, offset)

    def retrieve_behaviour(self, identityId: Uuid, reference: Uuid, config: Callable[[BehaviourReq], None]) -> \
            Observable[BehaviourRes]:
        return self.f_proc.retrieve_behaviour(identityId, reference, config)

    def retrieve_behaviour_executions(self, identity_id: Uuid, config: Callable[[BehaviourExecutionReq], None],
                                      limit: int = None, offset: int = None, startDate: str = None, endDate: str = None) \
            -> Observable[BehaviourExecutionRes]:
        return self.f_proc.retrieve_behaviour_executions(identity_id, config, limit, offset, startDate, endDate)

    def retrieve_behaviour_execution(self, identity_id: Uuid, process_instance_id: Uuid,
                                     config: Callable[[BehaviourExecutionDetailReq], None]) -> Observable[
        BehaviourExecutionDetailRes]:
        return self.f_proc.retrieve_behaviour_execution(identity_id, process_instance_id, config)

    def introspect(self, config: Callable[[IntrospectionReq], None]) -> Observable[IntrospectionRes]:
        return self.f_proc.introspect(config)

    def preserve(self, config: Callable[[PreservationReq], None]) -> Observable[PreservationRes]:
        return self.f_proc.preserve(config)

    def preserve_create_identities(self, impulses: List[CreateIdentityImpulse]) -> Observable[CreatedIdentityImpulse]:
        return self.f_proc.preserve_create_identities(impulses)

    def preserve_update_identities(self, impulses: List[UpdateIdentityImpulse]) -> Observable[UpdatedIdentityImpulse]:
        return self.f_proc.preserve_update_identities(impulses)

    def preserve_delete_identities(self, impulses: List[DeleteIdentityImpulse]) -> Observable[DeletedIdentityImpulse]:
        return self.f_proc.preserve_delete_identities(impulses)

    def preserve_create_intents(self, impulses: List[CreateIntentImpulse]) -> Observable[CreatedIntentImpulse]:
        return self.f_proc.preserve_create_intents(impulses)

    def preserve_update_intents(self, impulses: List[UpdateIntentImpulse]) -> Observable[UpdatedIntentImpulse]:
        return self.f_proc.preserve_update_intents(impulses)

    def preserve_delete_intents(self, impulses: List[DeleteIntentImpulse]) -> Observable[DeletedIntentImpulse]:
        return self.f_proc.preserve_delete_intents(impulses)

    def preserve_create_prompts(self, impulses: List[CreatePromptImpulse]) -> Observable[CreatedPromptImpulse]:
        return self.f_proc.preserve_create_prompts(impulses)

    def preserve_update_prompts(self, impulses: List[UpdatePromptImpulse]) -> Observable[UpdatedPromptImpulse]:
        return self.f_proc.preserve_update_prompts(impulses)

    def preserve_delete_prompts(self, impulses: List[DeletePromptImpulse]) -> Observable[DeletedPromptImpulse]:
        return self.f_proc.preserve_delete_prompts(impulses)

    def preserve_create_statements(self, impulses: List[CreateStatementImpulse]) -> Observable[CreatedStatementImpulse]:
        return self.f_proc.preserve_create_statements(impulses)

    def preserve_update_statements(self, impulses: List[UpdateStatementImpulse]) -> Observable[UpdatedStatementImpulse]:
        return self.f_proc.preserve_update_statements(impulses)

    def preserve_delete_statements(self, impulses: List[DeleteStatementImpulse]) -> Observable[DeletedStatementImpulse]:
        return self.f_proc.preserve_delete_statements(impulses)

    def preserve_create_fulfilments(self, impulses: List[CreateFulfilmentImpulse]) -> Observable[
        CreatedFulfilmentImpulse]:
        return self.f_proc.preserve_create_fulfilments(impulses)

    def preserve_update_fulfilments(self, impulses: List[UpdateFulfilmentImpulse]) -> Observable[
        UpdatedFulfilmentImpulse]:
        return self.f_proc.preserve_update_fulfilments(impulses)

    def preserve_delete_fulfilments(self, impulses: List[DeleteFulfilmentImpulse]) -> Observable[
        DeletedFulfilmentImpulse]:
        return self.f_proc.preserve_delete_fulfilments(impulses)

    def preserve_create_behaviours(self, impulses: List[CreateBehaviourImpulse]) -> Observable[CreatedBehaviourImpulse]:
        return self.f_proc.preserve_create_behaviours(impulses)

    def preserve_update_behaviours(self, impulses: List[UpdateBehaviourImpulse]) -> Observable[UpdatedBehaviourImpulse]:
        return self.f_proc.preserve_update_behaviours(impulses)

    def preserve_delete_behaviours(self, impulses: List[DeleteBehaviourImpulse]) -> Observable[DeletedBehaviourImpulse]:
        return self.f_proc.preserve_delete_behaviours(impulses)

    def preserve_create_codes(self, impulses: List[CreateCodeImpulse]) -> Observable[CreatedCodeImpulse]:
        return self.f_proc.preserve_create_codes(impulses)

    def preserve_update_codes(self, impulses: List[UpdateCodeImpulse]) -> Observable[UpdatedCodeImpulse]:
        return self.f_proc.preserve_update_codes(impulses)

    def preserve_delete_codes(self, impulses: List[DeleteCodeImpulse]) -> Observable[DeletedCodeImpulse]:
        return self.f_proc.preserve_delete_codes(impulses)

    def preserve_create_edges(self, impulses: List[CreateEdgeImpulse]) -> Observable[CreatedEdgeImpulse]:
        return self.f_proc.preserve_create_edges(impulses)

    def preserve_delete_edges(self, impulses: List[DeleteEdgeImpulse]) -> Observable[DeletedEdgeImpulse]:
        return self.f_proc.preserve_delete_edges(impulses)

    def perceive(self, config: Callable[[PerceptionReq], None]) -> Observable[PerceptionRes]:
        return self.f_proc.perceive(config)

    def perceive_action(self, impulse: PerceiveActionImpulse) -> Observable[PerceivedImpulse]:
        return self.f_proc.perceive_action(impulse)

    def perceive_data(self, impulse: PerceiveDataImpulse) -> Observable[PerceivedImpulse]:
        return self.f_proc.perceive_data(impulse)

    def practice(self, config: Callable[[PracticeReq], None]) -> Observable[PracticeRes]:
        return self.f_proc.practice(config)

    def practice_build(self, impulse: CreateSkillBuildJobImpulse,
                       config: Union[Callable[[CreatedSkillBuildJobImpulseReq], None], None] = None) -> Observable[
        CreatedSkillBuildJobImpulseRes]:
        return self.f_proc.practice_build(impulse, config)

    def practice_cancel(self, impulse: CancelSkillBuildJobImpulse,
                        config: Union[Callable[[CanceledSkillBuildJobImpulseReq], None], None] = None) -> Observable[
        CanceledSkillBuildJobImpulseRes]:
        return self.f_proc.practice_cancel(impulse, config)

    def introspect_build_jobs(self, tenant_id: Uuid, config: Union[Callable[[SkillBuildJobReq], None], None] = None) -> \
            Observable[SkillBuildJobRes]:
        return self.f_proc.introspect_build_jobs(tenant_id, config)

File: gaia-sdk-java/gaia-sdk-core/src/main/kotlin/gaia/sdk/core/Gaia.kt
package gaia.sdk.core

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import gaia.sdk.GaiaCredentials
import gaia.sdk.JWTCredentials
import gaia.sdk.Uuid
import gaia.sdk.api.ISensorFunction
import gaia.sdk.api.ISensorQueue
import gaia.sdk.api.ISensorStream
import gaia.sdk.http.HttpSensorFunction
import gaia.sdk.http.HttpSensorStream
import gaia.sdk.http.HttpTransportException
import gaia.sdk.http.TransporterFactory
import gaia.sdk.mqtt.MqttSensorQueue
import gaia.sdk.request.input.*
import gaia.sdk.request.type.BehaviourExecution
import gaia.sdk.request.type.BehaviourExecutionDetail
import gaia.sdk.request.type.Edge
import gaia.sdk.request.type.Experience
import gaia.sdk.request.type.Knowledge
import gaia.sdk.request.type.IdentityMetrics
import gaia.sdk.request.type.BehaviourMetrics
import gaia.sdk.request.type.Retrieval
import gaia.sdk.response.type.*
import gaia.sdk.spi.QueueOptions
import io.netty.buffer.Unpooled
import org.reactivestreams.Publisher
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.netty.http.client.HttpClient
import java.io.ByteArrayOutputStream

class Gaia {
    companion object {

        private val jsonparser = ObjectMapper().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        var transporterFactory = TransporterFactory()

        fun connect(url: String, credentials: GaiaCredentials): GaiaRef {
            return GaiaRef(GaiaConfig(url, credentials, transporterFactory))
        }

        fun connect(config: GaiaConfig): GaiaRef {
            return GaiaRef(config)
        }

        fun login(url: String, credentials: UsernamePasswordCredentials): GaiaCredentials {
            val loginRequest = jsonparser.writeValueAsBytes(credentials)
            val response = HttpClient.create()
                    .headers {
                        it.add("Content-Type", "application/json")
                    }
                    .followRedirect(true)
                    .post()
                    .uri("${url}/api/auth/access")
                    .send(Mono.just(Unpooled.copiedBuffer(loginRequest)))
                    .responseConnection { t, u ->
                        u
                                .inbound()
                                .receive()
                                .asByteArray()
                                .buffer()
                                .map { list ->
                                    val bos = ByteArrayOutputStream()
                                    list.forEach(bos::write)
                                    bos.toByteArray()
                                }
                                .switchIfEmpty(
                                        Flux.just(t.status())
                                                .flatMap {
                                                    if (it.code() >= 400) {
                                                        val msg = "Error with status code ${t.status().code()} (${t.status().reasonPhrase()}) and no payload"
                                                        Flux.error(HttpTransportException(msg))
                                                    } else {
                                                        Flux.empty<ByteArray>()
                                                    }
                                                }
                                )
                                .map { byteArray ->
                                    if (t.status().code() >= 400) {
                                        val msg = "Error with status code ${t.status().code()} (${t.status().reasonPhrase()}) and payload: ${String(byteArray)}"
                                        throw HttpTransportException(msg)
                                    }

                                    jsonparser.readValue(byteArray, LoginResponse::class.java)
                                }
                    }.cast(LoginResponse::class.java).blockFirst()

            response ?: throw Exception("No response received from login")

            return JWTCredentials(response.accessToken)
        }
    }
}


class GaiaConfig(val url: String,
                 val credentials: GaiaCredentials,
                 val transporterFactory: TransporterFactory,
                 val functionProcessor: ISensorFunction = HttpSensorFunction(url, credentials, transporterFactory),
                 val queueProcessor: ISensorQueue = MqttSensorQueue(QueueOptions("localhost", 1883)),
                 val streamProcessor: ISensorStream = HttpSensorStream(url, credentials, transporterFactory))

class GaiaRef(config: GaiaConfig) : ISensorFunction, ISensorStream {
    private val fProc: ISensorFunction = config.functionProcessor
    private val qProc: ISensorQueue = config.queueProcessor
    private val sProc: ISensorStream = config.streamProcessor

    override fun retrieve(config: Retrieval.() -> Unit) = fProc.retrieve(config)
    override fun retrieveExperience(config: Experience.() -> Unit) = fProc.retrieveExperience(config)
    override fun retrieveKnowledge(config: Knowledge.() -> Unit) = fProc.retrieveKnowledge(config)
    override fun retrieveEdges(source: Uuid, config: Edge.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveEdges(source, config, limit, offset)
    override fun retrieveEdge(source: Uuid, edgeId: Uuid, config: Edge.() -> Unit) = fProc.retrieveEdge(source, edgeId, config)
    override fun retrieveIntents(identityId: Uuid, config: gaia.sdk.request.type.Intent.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveIntents(identityId, config, limit, offset)
    override fun retrieveIntent(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Intent.() -> Unit) = fProc.retrieveIntent(identityId, reference, config)
    override fun retrieveIdentities(config: gaia.sdk.request.type.Identity.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveIdentities(config, limit, offset)
    override fun retrieveIdentity(identityId: Uuid, config: gaia.sdk.request.type.Identity.() -> Unit) = fProc.retrieveIdentity(identityId, config)
    override fun retrieveTenants(config: gaia.sdk.request.type.Tenant.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveTenants(config, limit, offset)
    override fun retrieveTenant(tenantId: Uuid, config: gaia.sdk.request.type.Tenant.() -> Unit) = fProc.retrieveTenant(tenantId, config)
    override fun retrieveUsers(config: gaia.sdk.request.type.User.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveUsers(config, limit, offset)
    override fun retrieveUser(userId: Uuid, config: gaia.sdk.request.type.User.() -> Unit) = fProc.retrieveUser(userId, config)
    override fun retrieveApiKeys(config: gaia.sdk.request.type.ApiKey.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveApiKeys(config, limit, offset)
    override fun retrieveApiKey(apiKeyId: Uuid, config: gaia.sdk.request.type.ApiKey.() -> Unit) = fProc.retrieveApiKey(apiKeyId, config)
    override fun retrieveRoles(tenantId: Uuid, config: gaia.sdk.request.type.Role.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveRoles(tenantId, config, limit, offset)
    override fun retrieveRole(tenantId: Uuid, roleId: Uuid, config: gaia.sdk.request.type.Role.() -> Unit) = fProc.retrieveRole(tenantId, roleId, config)
    override fun retrievePrompts(identityId: Uuid, config: gaia.sdk.request.type.Prompt.() -> Unit, limit: Int?, offset: Long?) = fProc.retrievePrompts(identityId, config, limit, offset)
    override fun retrievePrompt(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Prompt.() -> Unit) = fProc.retrievePrompt(identityId, reference, config)
    override fun retrieveStatements(identityId: Uuid, config: gaia.sdk.request.type.Statement.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveStatements(identityId, config, limit, offset)
    override fun retrieveStatement(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Statement.() -> Unit) = fProc.retrieveStatement(identityId, reference, config)
    override fun retrieveFulfilments(identityId: Uuid, config: gaia.sdk.request.type.Fulfilment.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveFulfilments(identityId, config, limit, offset)
    override fun retrieveFulfilment(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Fulfilment.() -> Unit) = fProc.retrieveFulfilment(identityId, reference, config)
    override fun retrieveCodes(identityId: Uuid, config: gaia.sdk.request.type.Code.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveCodes(identityId, config, limit, offset)
    override fun retrieveCode(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Code.() -> Unit) = fProc.retrieveCode(identityId, reference, config)
    override fun retrieveBehaviours(identityId: Uuid, config: gaia.sdk.request.type.Behaviour.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveBehaviours(identityId, config, limit, offset)
    override fun retrieveBehaviour(identityId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Behaviour.() -> Unit) = fProc.retrieveBehaviour(identityId, reference, config)
    override fun retrieveSkills(tenantId: Uuid, config: gaia.sdk.request.type.Skill.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveSkills(tenantId, config, limit, offset)
    override fun retrieveSkill(tenantId: Uuid, reference: Uuid, config: gaia.sdk.request.type.Skill.() -> Unit) = fProc.retrieveSkill(tenantId, reference, config)
    override fun retrieveSkillProvisions(tenantId: Uuid, config: gaia.sdk.request.type.SkillProvision.() -> Unit, limit: Int?, offset: Long?) = fProc.retrieveSkillProvisions(tenantId, config, limit, offset)
    override fun retrieveSkillProvision(tenantId: Uuid, reference: Uuid, config: gaia.sdk.request.type.SkillProvision.() -> Unit) = fProc.retrieveSkillProvision(tenantId, reference, config)
    override fun retrieveBehaviourExecution(identityId: Uuid, processInstanceId: Uuid, config: BehaviourExecutionDetail.() -> Unit) = fProc.retrieveBehaviourExecution(identityId, processInstanceId, config)
    override fun retrieveBehaviourExecutions(identityId: Uuid, config: BehaviourExecution.() -> Unit, limit: Int?, offset: Long?, startDate: String?, endDate: String?) = fProc.retrieveBehaviourExecutions(identityId, config, limit, offset, startDate, endDate)
    override fun retrieveIdentityMetrics(identityId: Uuid, startDate: String, endDate: String, config: IdentityMetrics.() -> Unit, limit: Int?) = fProc.retrieveIdentityMetrics(identityId, startDate, endDate, config, limit)
    override fun retrieveBehaviourMetrics(identityId: Uuid, behaviourId: Uuid?, startDate: String, endDate: String, config: BehaviourMetrics.() -> Unit, limit: Int?) = fProc.retrieveBehaviourMetrics(identityId, behaviourId, startDate, endDate, config, limit)
    override fun introspect(config: gaia.sdk.request.type.Introspection.() -> Unit) = fProc.introspect(config)
    override fun preserve(config: gaia.sdk.request.type.Preservation.() -> Unit) = fProc.preserve(config)
    override fun preserveCreateIdentities(vararg impulses: CreateIdentityImpulse) = fProc.preserveCreateIdentities(*impulses)
    override fun preserveUpdateIdentities(vararg impulses: UpdateIdentityImpulse) = fProc.preserveUpdateIdentities(*impulses)
    override fun preserveDeleteIdentities(vararg impulses: DeleteIdentityImpulse) = fProc.preserveDeleteIdentities(*impulses)
    override fun preserveCreateTenants(vararg impulses: CreateTenantImpulse): Publisher<CreatedTenantImpulse> = fProc.preserveCreateTenants(*impulses)
    override fun preserveUpdateTenants(vararg impulses: UpdateTenantImpulse): Publisher<UpdatedTenantImpulse> = fProc.preserveUpdateTenants(*impulses)
    override fun preserveDeleteTenants(vararg impulses: DeleteTenantImpulse): Publisher<DeletedTenantImpulse> = fProc.preserveDeleteTenants(*impulses)
    override fun preserveCreateUsers(vararg impulses: CreateUserImpulse): Publisher<CreatedUserImpulse> = fProc.preserveCreateUsers(*impulses)
    override fun preserveUpdateUsers(vararg impulses: UpdateUserImpulse): Publisher<UpdatedUserImpulse> = fProc.preserveUpdateUsers(*impulses)
    override fun preserveDeleteUsers(vararg impulses: DeleteUserImpulse): Publisher<DeletedUserImpulse> = fProc.preserveDeleteUsers(*impulses)
    override fun preserveCreateApiKeys(vararg impulses: CreateApiKeyImpulse): Publisher<CreatedApiKeyImpulse> = fProc.preserveCreateApiKeys(*impulses)
    override fun preserveUpdateApiKeys(vararg impulses: UpdateApiKeyImpulse): Publisher<UpdatedApiKeyImpulse> = fProc.preserveUpdateApiKeys(*impulses)
    override fun preserveDeleteApiKeys(vararg impulses: DeleteApiKeyImpulse): Publisher<DeletedApiKeyImpulse> = fProc.preserveDeleteApiKeys(*impulses)
    override fun preserveCreateRoles(vararg impulses: CreateRoleImpulse): Publisher<CreatedRoleImpulse> = fProc.preserveCreateRoles(*impulses)
    override fun preserveUpdateRoles(vararg impulses: UpdateRoleImpulse): Publisher<UpdatedRoleImpulse> = fProc.preserveUpdateRoles(*impulses)
    override fun preserveDeleteRoles(vararg impulses: DeleteRoleImpulse): Publisher<DeletedRoleImpulse> = fProc.preserveDeleteRoles(*impulses)
    override fun preserveCreateIntents(vararg impulses: CreateIntentImpulse) = fProc.preserveCreateIntents(*impulses)
    override fun preserveUpdateIntents(vararg impulses: UpdateIntentImpulse) = fProc.preserveUpdateIntents(*impulses)
    override fun preserveDeleteIntents(vararg impulses: DeleteIntentImpulse) = fProc.preserveDeleteIntents(*impulses)
    override fun preserveCreatePrompts(vararg impulses: CreatePromptImpulse) = fProc.preserveCreatePrompts(*impulses)
    override fun preserveUpdatePrompts(vararg impulses: UpdatePromptImpulse) = fProc.preserveUpdatePrompts(*impulses)
    override fun preserveDeletePrompts(vararg impulses: DeletePromptImpulse) = fProc.preserveDeletePrompts(*impulses)
    override fun preserveCreateStatements(vararg impulses: CreateStatementImpulse) = fProc.preserveCreateStatements(*impulses)
    override fun preserveUpdateStatements(vararg impulses: UpdateStatementImpulse) = fProc.preserveUpdateStatements(*impulses)
    override fun preserveDeleteStatements(vararg impulses: DeleteStatementImpulse) = fProc.preserveDeleteStatements(*impulses)
    override fun preserveCreateFulfilments(vararg impulses: CreateFulfilmentImpulse) = fProc.preserveCreateFulfilments(*impulses)
    override fun preserveUpdateFulfilments(vararg impulses: UpdateFulfilmentImpulse) = fProc.preserveUpdateFulfilments(*impulses)
    override fun preserveDeleteFulfilments(vararg impulses: DeleteFulfilmentImpulse) = fProc.preserveDeleteFulfilments(*impulses)
    override fun preserveCreateBehaviours(vararg impulses: CreateBehaviourImpulse) = fProc.preserveCreateBehaviours(*impulses)
    override fun preserveUpdateBehaviours(vararg impulses: UpdateBehaviourImpulse) = fProc.preserveUpdateBehaviours(*impulses)
    override fun preserveDeleteBehaviours(vararg impulses: DeleteBehaviourImpulse) = fProc.preserveDeleteBehaviours(*impulses)
    override fun preserveCreateCodes(vararg impulses: CreateCodeImpulse) = fProc.preserveCreateCodes(*impulses)
    override fun preserveUpdateCodes(vararg impulses: UpdateCodeImpulse) = fProc.preserveUpdateCodes(*impulses)
    override fun preserveDeleteCodes(vararg impulses: DeleteCodeImpulse) = fProc.preserveDeleteCodes(*impulses)
    override fun preserveCreateEdges(vararg impulses: CreateEdgeImpulse) = fProc.preserveCreateEdges(*impulses)
    override fun preserveDeleteEdges(vararg impulses: DeleteEdgeImpulse) = fProc.preserveDeleteEdges(*impulses)
    override fun preserveCreateSkills(vararg impulses: CreateSkillImpulse) = fProc.preserveCreateSkills(*impulses)
    override fun preserveUpdateSkills(vararg impulses: UpdateSkillImpulse) = fProc.preserveUpdateSkills(*impulses)
    override fun preserveDeleteSkills(vararg impulses: DeleteSkillImpulse) = fProc.preserveDeleteSkills(*impulses)
    override fun preserveCreateSkillProvisions(vararg impulses: CreateSkillProvisionImpulse) = fProc.preserveCreateSkillProvisions(*impulses)
    override fun preserveUpdateSkillProvisions(vararg impulses: UpdateSkillProvisionImpulse) = fProc.preserveUpdateSkillProvisions(*impulses)
    override fun preserveDeleteSkillProvisions(vararg impulses: DeleteSkillProvisionImpulse) = fProc.preserveDeleteSkillProvisions(*impulses)
    override fun preserveConnectNodeSet(nodeId: Uuid, impulse: ConnectSetNodeImpulse) = fProc.preserveConnectNodeSet(nodeId, impulse);
    override fun preserveConnectNodeUnset(nodeId: Uuid, impulse: ConnectUnsetNodeImpulse) = fProc.preserveConnectNodeUnset(nodeId, impulse);
    override fun preserveConnectNodeAppend(nodeId: Uuid, impulse: ConnectAppendNodeImpulse) = fProc.preserveConnectNodeAppend(nodeId, impulse);
    override fun preserveConnectNodeRemove(nodeId: Uuid, impulse: ConnectRemoveNodeImpulse) = fProc.preserveConnectNodeRemove(nodeId, impulse);
    override fun perceive(config: gaia.sdk.request.type.Perception.() -> Unit) = fProc.perceive(config)
    override fun perceiveAction(impulse: PerceiveActionImpulse) = fProc.perceiveAction(impulse)
    override fun perceiveData(impulse: PerceiveDataImpulse) = fProc.perceiveData(impulse)
    override fun introspectBuildJobs(tenantId: Uuid, config: (gaia.sdk.request.type.SkillBuildJob.() -> Unit)?): Publisher<SkillBuildJob> = fProc.introspectBuildJobs(tenantId, config)
    override fun practice(config: gaia.sdk.request.type.Practice.() -> Unit): Publisher<Practice> = fProc.practice(config)
    override fun practiceBuild(impulse: CreateSkillBuildJobImpulse, config: (gaia.sdk.request.type.CreatedSkillBuildJobImpulse.() -> Unit)?): Publisher<CreatedSkillBuildJobImpulse> =
            fProc.practiceBuild(impulse, config)
    override fun practiceCancel(impulse: CancelSkillBuildJobImpulse, config: (gaia.sdk.request.type.CanceledSkillBuildJobImpulse.() -> Unit)?): Publisher<CanceledSkillBuildJobImpulse> =
            fProc.practiceCancel(impulse, config)

    // data api
    override fun data(url: String) = sProc.data(url)
    // identity api
    override fun identity() = sProc.identity()
    // skill api
    override fun skill(url: String) = sProc.skill(url)
}

class UsernamePasswordCredentials(val username: String, val password: String)
data class LoginResponse @JsonCreator constructor(
        @JsonProperty("username") val username: String,
        @JsonProperty("accessToken") val accessToken: String
)
